docs/source/development/eep-01-pytrees.rst

```diff
@@ -0,0 +1,848 @@
+=======================
+
+All changes are fully backwards compatible.
+
+
+Motivation
+==========
+
+Estimagic has many functions that require user written functions as inputs. Examples
+are:
+
+- criterion functions and their derivatives for optimization
+- functions of which numerical derivatives are taken
+- functions that calculate simulated moments
+- functions that calculate bootstrap statistics
+
+In all cases, there are some restrictions on possible inputs and outputs of the
+user written functions. For example, parameters for numerical optimization need to
+be provided as pandas.DataFrame with a ``"value"`` column. Simulated moments and
+bootstrap statistics need to be returned as a pandas.Series, etc.
+
+Pytrees allow to relax many of those restrictions on interfaces of user provided
+functions. This is not only more convenient for users, but sometimes also allows to
+reduce overhead because the user can choose optimal data structures for their problem.
+
+
+Background: What is a pytree
+============================
+
+Pytree is a term used in TensorFlow and JAX to refer to a tree-like structure built out
+of container-like Python objects with arbitrary levels of nesting.
+
+What is a container can be re-defined for each application. By default, lists, tuples
+and dicts are considered containers and everything else is a leaf. Then the following
+are examples of pytrees:
+
+
+.. code-block:: python
+
+    [1, "a", np.arange(3)]  # 3 leaves
+
+    [1, {"k1": 2, "k2": (3, 4)}, 5]  # 5 leaves
+
+    np.arange(5)  # 1 leaf
+
+
+What makes pytrees so powerful are the operations defined for them. The most important
+ones are:
+
+- ``tree_flatten``: Convert any pytree into a flat list of leaves + metadata
+- ``tree_unflatten``: The inverse of ``tree_flatten``
+- ``tree_map``: Apply a function to all leaves in a pytree
+- ``leaf_names``: Generate a list of names for all leaves in a pytree
+
+The above examples of pytrees would look as follows when flattened (with a default
+definition of containers):
+
+.. code-block:: python
+
+    [1, "a", np.arange(3)]
+
+    [1, 2, 3, 4, 5]
+
+    [np.arange(5)]
+
+By adding numpy arrays to the registry of container like objects, each of the three
+examples above would have five leafs. The flattened versions would look as follows:
+
+.. code-block:: python
+
+    [1, "a", 0, 1, 2]
+
+    [1, 2, 3, 4, 5]
+
+    [0, 1, 2, 3, 4]
+
+Needless to say, it is possible to register any iterable as container. For example, we
+would add pandas.Series and pandas.DataFrame (with varying definitions, depending on the
+application).
+
+
+Difference between pytrees in JAX and estimagic
+===============================================
+
+Most JAX functions `only work with Pytrees of arrays
+<https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions>`_, i.e.
+pytrees where container types are dicts, lists and tuples and all leaves are arrays.
+
+There are two ways to look at such pytrees:
```

Comment:
The previous paragraph talked about Jax. I guess Jax takes the first route? 

*Aside: I would suggest avoiding the word "view" due to its typical meaning in the Python context, i.e., view vs. copy.*

Reply:
Replaced "view" by "perspective".